src/source-windivert.c

```diff
@@ -0,0 +1,682 @@
+static WinDivertQueueVars g_wd_qv[WINDIVERT_MAX_QUEUE];
+static uint16_t g_wd_num = 0;
+static SCMutex g_wd_init_lock;
+
+void *WinDivertGetThread(int n)
+{
+    if (n >= g_wd_num) {
+        return NULL;
+    }
+    return (void *)&g_wd_tv[n];
+}
+
+void *WinDivertGetQueue(int n)
+{
+    if (n >= g_wd_num) {
+        return NULL;
+    }
+    return (void *)&g_wd_qv[n];
+}
+
+/**
+ * \brief initialize a WinDivert filter
+ *
+ * \param filter a WinDivert filter string as defined at
+ * https://www.reqrypt.org/windivert-doc.html#filter_language
+ *
+ * \retval 0 on success
+ * \retval -1 on failure
+ */
+int WinDivertRegisterQueue(bool forward, char *filter_str)
+{
+    SCEnter();
+    int ret = 0;
+
+    WINDIVERT_LAYER layer =
+            forward ? WINDIVERT_LAYER_NETWORK_FORWARD : WINDIVERT_LAYER_NETWORK;
+
+    /* validate the filter string */
+    const char *error_str;
+    uint32_t error_pos;
+    bool valid = WinDivertHelperCheckFilter(
+            filter_str, layer, &error_str, &error_pos);
+    if (!valid) {
+        SCLogWarning(SC_ERR_WINDIVERT_INVALID_FILTER,
+                "Invalid filter \"%s\" supplied to WinDivert: %s at position "
+                "%" PRId32 "",
+                filter_str, error_str, error_pos);
+        SCReturnInt(SC_ERR_WINDIVERT_INVALID_FILTER);
+    }
+
+    /* initialize the queue */
+    SCMutexLock(&g_wd_init_lock);
+
+    if (g_wd_num >= WINDIVERT_MAX_QUEUE) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT,
+                "Too many WinDivert queues specified %" PRId32 "", g_wd_num);
+        ret = -1;
+        goto unlock;
+    }
+    if (g_wd_num == 0) {
+        /* on first registration, zero-initialize all array structs */
+        memset(&g_wd_tv, 0, sizeof(g_wd_tv));
+        memset(&g_wd_qv, 0, sizeof(g_wd_qv));
+    }
+
+    /* init thread vars */
+    WinDivertThreadVars *wd_tv = &g_wd_tv[g_wd_num];
+    wd_tv->thread_num = g_wd_num;
+
+    /* init queue vars */
+    WinDivertQueueVars *wd_qv = &g_wd_qv[g_wd_num];
+    wd_qv->queue_num = g_wd_num;
+
+    /* copy filter to persistent storage */
+    size_t filter_len = strlen(filter_str);
+    size_t copy_len =
+            strlcpy(wd_qv->filter_str, filter_str, sizeof(wd_qv->filter_str));
+    if (filter_len > copy_len) {
+        SCLogWarning(SC_ERR_WINDIVERT_TOOLONG_FILTER,
+                "Queue length exceeds storage by %" PRId32 " bytes",
+                (int32_t)(filter_len - copy_len));
+        ret = -1;
+        goto unlock;
+    }
+
+    wd_qv->layer = layer;
+    wd_qv->priority =
+            g_wd_num; /* priority set in the order filters are defined */
+    wd_qv->flags = 0; /* normal inline function */
+
+    SCMutexInit(&wd_qv->filter_init_mutex, NULL);
+    SCMutexInit(&wd_qv->counters_mutex, NULL);
+
+    g_wd_num++;
+
+unlock:
+    SCMutexUnlock(&g_wd_init_lock);
+
+    if (ret == 0) {
+        /* not really a "device" identical to nfq/ipfw, but best descriptor */
+        LiveRegisterDevice(filter_str);
```

Comment:
have you seen this value being used anywhere? In other runmodes it may be used in thread names and unix socket. The latter is obviously irrelevant here. Not sure about the former.

Reply:
Thread names. I'll do what NFQ does, which is just use the queue number (in this case I'll use the array index).